### ISSUE TITLE
Different LocalDate values inserted with H2

### ISSUE THREAD

**Issue by npostnikova** [jOOQ/jOOQ#13506]
I generate a [Task table](https://github.com/bardsoftware/ganttproject/blob/master/ganttproject/src/main/resources/resources/sql/init-project-database.sql#L9) with a `start_date` column of type `date`.
Here's the [jOOQ generator](https://github.com/bardsoftware/ganttproject/blob/0488c54b8d5ba96b549bdcc8dd255b9ededf231f/ganttproject/build.gradle#L68). The DSL accepts `LocalDate` type for the field.
It turned out that executing the insert statement generated by jOOQ differs from executing its string representation. Moreover, the latter looks more reasonable :)

### Expected behavior
The same value of the `start_date` column in the cases below, i.e. when printing the fetch result. 
Case 1:
```
val insertQuery = buildInsertTaskQuery(dsl, task)
insertQuery.execute()
```
Case 2:
```
dsl.execute(insertQuery.getSQL(ParamType.INLINED))
```

### Actual behavior
In case 1, the day before is inserted.

Case 1:
```
buildInsertTaskQuery(dsl, task).execute()
println(task.start.toLocalDate())  // 2023-01-06  -- the value passed to jOOQ
println(dsl.selectFrom(TASK).fetch()[0]) // 2023-01-05 -- the value held in the column
```
Case 2:
```
dsl.execute(buildInsertTaskQuery(dsl, task).getSQL(ParamType.INLINED))
println(task.start.toLocalDate())  // 2023-01-06 -- the value passed to jOOQ
println(dsl.selectFrom(TASK).fetch()[0]) // 2023-01-06  -- the value held in the column
```
### Steps to reproduce the problem

It works for me this way if I perform the actions described in [this test case](https://github.com/bardsoftware/ganttproject/blob/master/ganttproject-tester/test/net/sourceforge/ganttproject/storage/ProjectDatabaseTest.kt#L101) environment. I mean removing everything starting with `projectDatabase.insertTask(task)` and placing one of the cases instead.

### Versions

- jOOQ: 7.1.+
- Java: Oracle OpenJDK 17.0.2
- Database (include vendor): H2
- OS: Ubuntu 20.04.4 LTS
- JDBC Driver (include name if inofficial driver): com.h2database 2.1.+

**Comment by lukaseder** [comment]
Thanks for your report.

I'm nto aware of such a bug. Our integration tests don't hint at this happening. In order to better isolate the problem (I'd have to browse through your code to find how you set up your JVM, JDBC connection, etc.), would it be possible to create a more minimal example? We have a template here: https://github.com/jOOQ/jOOQ-mcve

**Comment by npostnikova** [comment]
Hey, I wasn't able to reproduce it using the template yet. I'll let you know if there're any updates, even if it was a false alarm :)

**Comment by lukaseder** [comment]
Thanjs, yes will be good to know what it was, irrespective of whether it's a bug or setup issue.

I'm assuming it really is related to differing time zones somewhere.

**Comment by katzyn** [comment]
If time zones of connection and JVM are out of sync, such issues may appear with legacy defective by design `java.sql.Date`, `java.sql.Time`, and `java.sql.Timestamp` classes.

For example,
```Java
TimeZone.setDefault(TimeZone.getTimeZone("GMT+01"));
// Time zone isn't specified as a setting, so current default will be used by this connection
Connection c = DriverManager.getConnection("jdbc:h2:mem:");

TimeZone.setDefault(TimeZone.getTimeZone("GMT+02"));
PreparedStatement ps = c.prepareStatement("SELECT CAST(? AS DATE)");
ps.setDate(1, java.sql.Date.valueOf(LocalDate.of(2023, 1, 6)));
ResultSet rs = ps.executeQuery();
rs.next();
System.out.println(rs.getObject(1, LocalDate.class));
```
prints 2023-01-05.

Actually you can run into such issue even without databases, the following code
```Java
TimeZone.setDefault(TimeZone.getTimeZone("GMT+02"));
java.sql.Date date = java.sql.Date.valueOf(LocalDate.of(2023, 1, 6));
TimeZone.setDefault(TimeZone.getTimeZone("GMT+01"));
System.out.println(date);
````
also prints 2023-01-05.

**Comment by npostnikova** [comment]
@katzyn, thank you!
Also, I wanted to attach [this issue](https://github.com/h2database/h2database/issues/2573) as I suspected that your responses there describe the current case:

> java.time.LocalDate should always be passed directly to H2 and read directly from H2 ...
> If some implementation of JPA uses legacy java.sql.Date in the middle, this implementation should be fixed instead

**Comment by lukaseder** [comment]
I couldn't reproduce this particular issue at the time. However, there's been a recent fix related to DST:

- https://github.com/jOOQ/jOOQ/issues/14866

The fix was included in 3.20.0, 3.19.7, 3.18.14, 3.17.23. I'm closing this as "works for me," for now. Happy to reopen if there are instructions to reproduce the issue in an MCVE, or similar.
